Add popup with a report to list all cards

pmanrubia opened the pull request:
Paired with @leenagupte 
- [Trello card](https://trello.com/c/Y2u4rX5Y/192-custom-trello-report-via-api)
### Need

It is proving a bit difficult to manage all the cards that currently are in 'With department' or 'Out for factcheck' columns. It's a laborious manual process to chase departments and agencies, and we'd like to make it easier.
One way to do this is to build a custom report that queries the Trello API to create a spreadsheet containing details of all tickets in those two states.
This report can then be sent to departmental SPOCs on a weekly basis to inform them what's in the danger zone - that is, what might be closed if we don't hear back from them in reasonable time.

**Required columns**
- GDS Ref no. (Trello card ID) - hyperlink to the relevant Trello card
- Zendesk ID - hyperlink to the relevant Zendesk ticket
- Publishing URL
- Card name
- Description
- Labels
- Status (current column)
- How long (days)? (how long the card has been in its current status)
### Not Included
- Creating Excel Hyperlinks  (clickable name that points to URL)
- Need to be styled to Gov.UK standard

pmanrubia committed:
Move columnHeading closer to where it is used.

pmanrubia committed:
Move `$excel_btn` variable closer to where it is used.

pmanrubia committed:
Move code related to configure the menu item

This is done in preparation to extract an object responsible with
 creating the menu configuration.

pmanrubia committed:
Extract Menu class to configure menu

Wrap all menu code in its own class so it can be tested and configured
independently.

pmanrubia committed:
Move init class to the top

It improves readability because this is the public interface.

pmanrubia committed:
Apply convention of '_' for private methods

pmanrubia committed:
Convert variable `add_interval` into Object property

pmanrubia committed:
Rename`add_interval` to `_add_interval`

 Improves readability as this property is not exposed for other
 collaborations

pmanrubia committed:
Extract variable to remove comment

Use variable to make redundant a comment.

pmanrubia committed:
Convert function `_addExportLink` into method

pmanrubia committed:
Extract method `_optionAlreadyExist` and remove comment

pmanrubia committed:
Extract method to add the Export option to the menu

pmanrubia committed:
Extract method to check if the overlay is opened

pmanrubia committed:
Extract function to create an Excel button

pmanrubia committed:
Add missing variable

pmanrubia committed:
Re-indent manifest.json to 2 spaces

pmanrubia committed:
Move menu.js to src folder

pmanrubia committed:
Inline variable `i` so it is clear where it is used.

pmanrubia committed:
Move var declaration closer in variable: bytestring

pmanrubia committed:
Move var declaration closer in variable: buffer

pmanrubia committed:
Move var declaration closer in variable: ia

pmanrubia committed:
Remove points from the Excel report

pmanrubia committed:
Extract Excel Class to create the file.

pmanrubia committed:
Move memberIDs closer to where its used.

pmanrubia committed:
Move memberInitials closer to where its used.

pmanrubia committed:
Move labels closer to where its used.

pmanrubia committed:
Move rowData closer to where its used.

pmanrubia committed:
Move dueDate closer to where its used.

pmanrubia committed:
Initialize duedate for when exporting the card

pmanrubia committed:
Make variable names explicit

pmanrubia committed:
Move rowNum(s) variables closer to where they are used

pmanrubia committed:
Inline the card name

We have removed the `archived` title because it is not relevant for us
and simplifies the code.

pmanrubia committed:
Inline the list name

We have removed the `archived` title because it is not relevant for us
and simplifies the code.

pmanrubia committed:
Remove worksheet of archived cards

We don’t want the data from archived cards, so this tab has been removed.

pmanrubia committed:
Inline columnHeadings 

After removing the archived worksheet it is only used once.

pmanrubia committed:
Rename variable into meaningful name

pmanrubia committed:
Simplify how a row is added to the worksheet

It seems that is pushing an empty array is not needed to add a new row

pmanrubia committed:
Extract SpreadSheet class to set up the spreadSheet

It allows us to add functions to addRows, and do the initial setup.

pmanrubia committed:
Move export function into SpreadSheet class

They seem to be related as they share state (file object)

pmanrubia committed:
Rename excel.js to spreadSheet.js

pmanrubia committed:
Remove members info from the SpreadSheet

pmanrubia committed:
Make header configurable

The SpreadSheet does not depend on the header

pmanrubia committed:
Remove global variable declaration

pmanrubia committed:
Remove list column

We don’t need to iterate through the lists because we are only going
to use one.

pmanrubia committed:
Add all rows at once from the main function

This is done in preparation for extracting the logic that transform the
data from Trello to the SpreadSheet.

pmanrubia committed:
Extract function with the logic to transform Trello Cards

This is done in preparation to decide how we model into objects and methods
this functionality.

pmanrubia committed:
Extract Factcheck class

Initially the Factcheck class is just to move the code that transforms 
Trello cards into spreadsheet rows for any Trello board

The aim of this class is to eventually be specific to the Factcheck 
column.

pmanrubia committed:
Move Factcheck class into report folder

pmanrubia committed:
Extract TrelloAPI class to get data form Trello

pmanrubia committed:
Move TrelloAPI class into src folder

pmanrubia committed:
Move close button out of the call to the Trello API

pmanrubia committed:
Move call to get all Cards to the Factcheck class

pmanrubia committed:
Pass only the data that Factcheck uses

pmanrubia committed:
Do not close the popup menu

pmanrubia committed:
Move create function into menu configurable option

pmanrubia committed:
Rename createExcelReport to toSpreadSheet

pmanrubia committed:
Add Jasmine support

pmanrubia committed:
Add test to #addHeader

We convert the variable worksheet into a property in order to expose 
it to the test. This is something to be improved in forthcoming commits.

pmanrubia committed:
Add test to #addRows

pmanrubia committed:
Make addRow a private method

pmanrubia committed:
Remove additional empty list in worksheet set up

This second list was being added as a placeholder for the Archive tab
which we have previously removed.

leenagupte committed:
Add test to ensure that the spreadsheet is created with the expected name

pmanrubia committed:
Add constructor to SpreadSheet

pmanrubia committed:
Add test to ensure that the spreadsheet has content

pmanrubia committed:
SpreadSheet minor reformat

pmanrubia committed:
Update spread sheet author name

pmanrubia committed:
Minor: reformat spreadSheet

pmanrubia committed:
Test data is requested for the open Trello board

pmanrubia committed:
Test we get the board name when retrieving all cards

Changed to use an `ajax` request rather than getJSON to call the Trello API.
Using `ajax` allows us to mock the promise (done function) and “resolve” 
what we expect to receive from Trello. This allows us to test what is
being passed to the callback.

pmanrubia committed:
Test we get all cards from Trello

pmanrubia committed:
Make the transformRows function private

pmanrubia committed:
Extract method to process the row data

This facilitates testing as we don’t have to stub the asynchronous 
request.

pmanrubia committed:
Remove constructor function and use Object

This facilitates testing as it is easier to spy on this unique object
when it exist in a callback.

pmanrubia committed:
Add test for Factcheck#toSpreadSheet

pmanrubia committed:
Add a method to get all rows from the worksheet

We are assuming that there is only ever one worksheet in a spreadsheet.

Add a method to get all of the rows encapsulate how we store the rows.

pmanrubia committed:
Minor: add some missing ;

pmanrubia committed:
Add test all cards are added to spreadsheet

Created a JSON object to hold the board data. We tried using a fixture
file using jasmine-jquery.js[1], but were unable to get past the csrf 
errors raised by Chrome.


[1]: https://github.com/velesin/jasmine-jquery

pmanrubia committed:
Move test of #getRows to the top of the suite

Moved because this function is being used later on.

pmanrubia committed:
Use #getRows() method to avoid accessing the spreadSheet

pmanrubia committed:
Test that the spread sheet has a header

pmanrubia committed:
Test spreadsheet has the card name

pmanrubia committed:
Add a method to a single row in the spreadsheet

pmanrubia committed:
Use #getRow instead of reading directly from array

pmanrubia committed:
Add a test for the card description

pmanrubia committed:
Add test for the card ID

pmanrubia committed:
Add test for the card URL

pmanrubia committed:
Add test for the card status

pmanrubia committed:
Update Trello API URL to get card attachments

Also simplify the URL, by not requesting items that we don’t need.

pmanrubia committed:
Get the Zendesk ticket URL from the card attachments

pmanrubia committed:
Get the Zendesk ticket ID from the card attachments

pmanrubia committed:
Add Department / Agency to the report

pmanrubia committed:
Fix lint errors and reformatting

leenagupte committed:
Get actions from the Trello API

The actions hold data about the movement of cards. When a card is moved
to another list, the board actions are updated.

Use the action to get the date a card was moved to a list, and to
calculate the number of days it has been since the card was moved.

The list name is currently hard-coded to “List One” for the tests. This
will be updated to Factcheck in forthcoming commits

leenagupte committed:
Add moment.js library to calculate distance between dates

This is a preparatory commit because as per requirements we need to
display how long a card has been placed in the factcheck column

pmanrubia committed:
Remove 0's from the Excel file

Empty strings are currently converted into 0s when the excel file is
generated. This is a quick solution that displays a `-` instead

pmanrubia committed:
Add publishing URL to the excel report

pmanrubia committed:
Filter actions that represent a change of list

It also increases the number of actions to 1000, as before, with 50 we
were not able to calculate the time on the list

pmanrubia committed:
Renames TrelloAPI into Trello

This is a prepatory commit. The intention is to move all the logic that
need to know about the data returned by Trello into Trello.js. ie. search 
for a Zendesk attachment, the days in the list…

This will also facilitate reusing in other reports, although this is not
what is driving this change.

pmanrubia committed:
Move findZendeskId method to Trello

Testing is now much easier. The unit test only involves the card, and
the report test can rely on a Spy on the Trello object because it is
already tested.

I would need to get rid of the fixture, but as it is used in other tests
I will wait until all the code has been moved there.

pmanrubia committed:
Move findZendeskURL method to Trello

Testing is now much easier. The unit test only involves the card, and
the report test can rely on a Spy on the Trello object because it is
already tested.

I would need to get rid of the fixture, but as it is used in other tests
I will wait until all the code has been moved there.

pmanrubia committed:
Move findPublishingURL method to Trello

Testing is now much easier. The unit test only involves the card, and
the report test can rely on a Spy on the Trello object because it is
already tested.

I would need to get rid of the fixture, but as it is used in other tests
I will wait until all the code has been moved there.

pmanrubia committed:
Move findCardStatusDays method to Trello

Testing is now much easier. The unit test only involves the card, and
the report test can rely on a Spy on the Trello object because it is
already tested.

I would need to get rid of the fixture, but as it is used in other tests
I will wait until all the code has been moved there.

pmanrubia committed:
Delete unused fixtures

As per previous commits, once the unit testing is being done in the
Trello class, there is no need to big fixtures.

pmanrubia committed:
Renames fixture from list to column

This is more consistent with Trello use.

pmanrubia committed:
Move findLabel method to Trello

Testing is now much easier. The unit test only involves the card, and
the report test can rely on a Spy on the Trello object because it is
already tested.

I would need to get rid of the fixture, but as it is used in other tests
I will wait until all the code has been moved there.

pmanrubia committed:
Dry-up Factcheck

pmanrubia committed:
Dry-up Trello

pmanrubia committed:
Renames method to calculate the days in a column

pmanrubia committed:
Rename fixture

pmanrubia committed:
Dry-up SpreadSheet

pmanrubia committed:
Declare Trello methods independently

pmanrubia committed:
Mixup dryup factcheck

pmanrubia committed:
Move boardUrl out of Trello class

This is a preparatory commit to make accessible to the report to a
pop-up menu in the Chrome extension.

The Chrome extension pop-up run the Factcheck report in its own html.
For this reason `window.location.pathname` is not returning the board
url, but the popup url. 

By extracting this information out of the Trello object we prepare
next popup so it can pass the URL as a parameter.

pmanrubia committed:
Use url instead of path to extract the board id

As a continuation of previous commit, the Chrome extension will pass to
the Trello object the URL of the current tab, not its path.

pmanrubia committed:
Add popup menu to run the report

Remove the additional option for Trello menu, and add a popup with a
link to generate the report.

I find it much easier to use.

pmanrubia committed:
Minor refactorings

pmanrubia committed:
Update README.md

pmanrubia committed:
Add the column name to the SpreadSheet

Add the list as a 4th parameter to the callback. It might be interesting
to wrap this for parameters into an object with 4 attributes.

pmanrubia committed:
Renames FactCheck into AllCards

The report is no longer restricted to the FactCheck column, so the name
need to be updated. As it is currently listing all cards for all columns
I have renamed the name to AllCards

leenagupte commented:
@pmanrubia 👍 👍 

Only one teensy tiny thing. In the spreadsheet, I would change the Card ID field to return the short link.

So, for card https://trello.com/c/fJUBMPkr/245-consider-adding-option-to-gov-uk-chrome-extension-to-view-govspeak-version-of-page,
rather than seeing an id of "5799e9f598ba5d2d6b35ebbb", it'll show an id of "fJUBMPkr".

in the Trello API, the shortlink is returned as `shortLink`.

leenagupte committed:
Update icon image

leenagupte committed:
Update README with where to find the zip on GitHub

leenagupte commented:
@pmanrubia I added a couple of things:
1. Update the icon image, as agreed :)
2. Added a few lines to the README about how to download the zip.

pmanrubia committed:
Display shortLink instead of card id

The card ID is a mongo ID Value which is useless for the user.

pmanrubia commented:
@leenagupte: Thank you for your last commits. 
I have address your last comment in https://github.com/alphagov/trello-tracker/pull/1/commits/a5a5f76d2d42de65f476f4aef57cf54883d10101

Moving for fact-check

pmanrubia committed:
Rename column name that describe `time in a column`

Address feedback in https://trello.com/c/Y2u4rX5Y/192-custom-trello-report-via-api during the factcheck process

pmanrubia committed:
Use the card shortId as the ID for the report

The idShort is the number that the Trello Extension to display the card numbers
uses: https://chrome.google.com/webstore/detail/trello-card-numbers/kadpkdielickimifpinkknemjdipghaf?hl=en

pmanrubia committed:
Use Excel function to calculate time in a column

We no longer need moment.js
